src/lib.rs

```diff
@@ -343,6 +347,136 @@ impl Invariant {
         self.positions.get_all(&owner_id)
     }
 
+    pub async fn swap(
+        &mut self,
+        pool_key: PoolKey,
+        x_to_y: bool,
+        amount: TokenAmount,
+        by_amount_in: bool,
+        sqrt_price_limit: SqrtPrice,
+    ) -> Result<CalculateSwapResult, InvariantError> {
+        if exec::gas_available() < 25000000 * 2 {
```

Comment:
In the swap entrypoint, there is a different estimation depending on how liquidity is spread. The amount of gas depends on the number of iterations in the swap (swap step). Checking the gas amount should be done after `calculate_swap_result` then, the gas amount used to perform state update and transfer tokens are quite deterministic